### PR TITLE
Error handling and logging

### LINKED ISSUE #631: Breaking change: null data returned when resource not found
## Description
Earlier, when a resource does not exist, a 404 with an empty response body was returned. After fetch of latest master (with v4 merged in), I am now getting 404 with the next body:

Request URL: https://localhost/api/users/123

Response body:
```json
{
  "links": {
    "self": "https://localhost/api/users"
  },
  "data": null
}
```

Note the `self` link is wrong too: it misses the identity.

Traced this down to https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/a801d958d9640365f459e096b802ad5d90db33fa/src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs#L108, so it may be intentional.

### LINKED ISSUE #638: JsonApiException namespace issue
The current `JsonApiException` is situated in the `JsonApiDotNetCore.Internal` namespace, it should be in the `JsonApiDotNetCore.Internal.Exceptions`, as we have a folder for that.

As discussed in #620, where we discussed that namespaces are a bit of a mess at the moment, we should take some time to correct this.

### LINKED ISSUE #655: Invalid Patch request body format required (most likely user error)
## Description
Either I am doing something incorrectly (likely), or there is an inconsistency with PATCH request required format (unlikely).  I also realize that I am operating with an alpha version, and as such things might be goofy.

I am having a few issues, but the salient issue for this post is that I can't get a PATCH request to result in a deserialized type with the format specified in the json:api spec 1.0

Expected format per spec:
`{
	"data": {
		"type": "users",
		"id": 10,
		"attributes": {
			"userType": "Employee"
		}
	}
}`
This format actually results in a null type being presented to the JsonApiController PatchAsync method.

Actual format required to get a deserialized type that contains the value presented in the userType property (:
`{
    "type": "users",
    "id": "10",
    "userType": "Employee"
}`

All other imagined variations of the request body format either result in null types or types with default values that do not reflect the values passed in the request body.
...

## Environment

- Targeted Version: .netcore3.0
- JsonApiDotNetCore Version: 4.0.0-alpha4
- Other Relevant Package Versions:
-- Microsoft.AspNetCore.Mvc.NewtonsoftJson: 3.1.0
-- Newtonsoft.Json: 12.0.3

### LINKED ISSUE #689: Review logging
## Description
Now that we have the right logging dependencies in place, we should review when, what and at which level logging occurs within JADNC.

Official guidance on logging levels: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-2.2#log-level

I'm thinking of:
- Do not log events at **error** level which cannot be solved by us, but are functional errors that do not require investigation. Like resource-not-found, invalid-attribute-name, resource-already-exists, input validation errors etc.
- Basic trace/debug logging of JSON request/response, along with qualified URL
- Log query string parsing results, maybe

Additions are welcome!

### LINKED ISSUE #709: Bug: crash on invalid relationship name
When getting a misspelled relationship name, the code crashes at [CurrentRequestMiddleware.GetCurrentEntity()](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/ec5f8d560209738288f8907d37d3b319cfdd34e7/src/JsonApiDotNetCore/Middleware/CurrentRequestMiddleware.cs#L232) due to recent changes.

Repro steps:
- Run the unit test `Can_Get_TodoItem_WithOwner` to have a record in database
- Start project JsonApiDotNetCoreExample
- GET http://localhost:1479/api/v1/todoItems/1?include=owner and see todoItems resource followed by included owner
- GET http://localhost:1479/api/v1/todoItems/1/owner and observe successful response with person
- GET http://localhost:1479/api/v1/todoItems/1/ownerX and observe HTTP 500

### PR DESCRIPTION
**Pull request by bart-degreed**
This PR is mostly about error handling and logging.

- Ensured all errors go through an extensibility point (`IExceptionHandler`), which makes logging and exception handling customizable (see `CustomExceptionHandler` in tests). `DefaultExceptionHandler` provides virtual methods to adapt logging level, log message and the conversion from Exception to Error response objects.
- Converted action results into rich `JsonApiException` objects and added/updated tests to assert on the returned error objects.
- Fixed broken model state validation.
- Fixes around returning response body with NotFound for resources and relationships.
- Improved selection for which parser handles a query string parameter, allowing for better error messages.
- Fixes on handling of omitNull/omitDefault query string overrides.
- Improvements for handling `DisableQueryAttribute`.
- Added trace logging on controller/service/repository flow + JSON requests/responses.
- Fix in handling zero page size; set default page size to nonzero.

Fixes #631
Fixes #689
Fixes #655
Fixes #709
Fixes #638

### CODE REVIEW

File: src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs [hunk 1]
```diff
@@ -61,107 +61,97 @@ public abstract class BaseJsonApiController<T, TId> : JsonApiControllerMixin whe
...
-                return UnprocessableEntity();
+                throw new InvalidRequestBodyException(null, null, null);
 
             if (!_jsonApiOptions.AllowClientGeneratedIds && !string.IsNullOrEmpty(entity.StringId))
-                return Forbidden();
+                throw new ResourceIdInPostRequestNotAllowedException();
 
             if (_jsonApiOptions.ValidateModelState && !ModelState.IsValid)
-                return UnprocessableEntity(ModelState.ConvertToErrorCollection<T>());
+                throw new InvalidModelStateException(ModelState, typeof(T), _jsonApiOptions);
```
Comment (maurei) [inline src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs:117]: `JsonApiOptions` is used for only one of its properties. Perhaps we can remove it as a dependency and only pass a boolean to reflect this option?

File: src/JsonApiDotNetCore/Middleware/ConvertEmptyActionResultFilter.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+
+namespace JsonApiDotNetCore.Middleware
+{
+    public sealed class ConvertEmptyActionResultFilter : IAlwaysRunResultFilter
```
Comment (maurei) [inline src/JsonApiDotNetCore/Middleware/ConvertEmptyActionResultFilter.cs:7]: 👍

File: src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs (hunk 1 above)
Comment (bart-degreed) [reply src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs:117]: How to inject a boolean property?

Comment (bart-degreed) [reply src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs:117]: Update: I'm going to merge this, so I can work on related things. Will address this later.

Comment (bart-degreed) [reply src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs:117]: Fixed in #716.

**Pull request merged**